File: merge_sort_step.java
import utils.CsvReader;
import utils.CsvReader.DataPair;
import java.util.List;
import java.util.ArrayList;
import java.io.BufferedWriter;
import java.io.FileWriter;
import java.io.IOException;
import java.util.Scanner;

public class merge_sort_step extends CsvReader {

	public List<DataPair> MergeSortWithSteps(List<DataPair> data, List<List<DataPair>> snapshots){
		if (data.size() <= 1){
			return data;
		}
		int mid = (data.size()+1)/2;

		List<DataPair> left = MergeSortWithSteps(data.subList(0, mid), snapshots);

		List<DataPair> right = MergeSortWithSteps(data.subList(mid, data.size()), snapshots);
		// System.out.println("Left Tree " + count);

		List<DataPair> merged = merge(left, right, snapshots);
		// snapshots.add(new ArrayList<>(merged));
		return merged;
	}
	public void mergeSortInPlace(List<DataPair> list, List<DataPair> aux, int left, int right, List<List<DataPair>> snapshots) {
		if (right - left <= 1) return; // base case

		int mid = (left + right + 1) / 2;

		mergeSortInPlace(list, aux, left, mid, snapshots);
		mergeSortInPlace(list, aux, mid, right, snapshots);

		mergeAndSnapshot(list, aux, left, mid, right, snapshots);
	}


	private void mergeAndSnapshot(List<DataPair> list, List<DataPair> aux, int left, int mid, int right, List<List<DataPair>> snapshots) {
		int i = left, j = mid, k = left;

		while (i < mid && j < right) {
			if (list.get(i).id <= list.get(j).id) {
				aux.set(k++, list.get(i++));
			} else {
				aux.set(k++, list.get(j++));
			}
		}
		while (i < mid) aux.set(k++, list.get(i++));
		while (j < right) aux.set(k++, list.get(j++));

		for (int t = left; t < right; t++) {
			list.set(t, aux.get(t)); // copy back into original
		}

		snapshots.add(new ArrayList<>(list));
	}

	public List<DataPair> merge(List<DataPair> left, List<DataPair> right, List<List<DataPair>> snapshots){
		List<DataPair> merged = new ArrayList<>();
		int i = 0, j = 0;
		while (i < left.size() && j < right.size()){
			if (left.get(i).id <= right.get(j).id){
				merged.add(left.get(i));
				i++;
			} else {
				merged.add(right.get(j));
				j++;
			}
		}
		while (i < left.size()){
			merged.add(left.get(i));
			i++;
		}
		while (j < right.size()){
			merged.add(right.get(j));
			j++;
		}
		snapshots.add(new ArrayList<>(merged));
		return merged;
	}
	public static void main(String[] args) {
		CsvReader reader = new CsvReader();
		
		List<DataPair> data = null;
		List<DataPair> aux = null;
		int startingRow = 0;
		int endingRow = 0;
		try (Scanner scanner = new Scanner(System.in)) {
			System.out.print("Enter the path to the CSV file: ");
			String csvFile = scanner.nextLine();
			data = reader.readCsv(csvFile);
			System.out.print("Enter the starting row: ");
			startingRow = scanner.nextInt();
			System.out.print("Enter the ending row: ");
			endingRow = scanner.nextInt();
			data = reader.readCsvWithStartAndEnd(csvFile, startingRow, endingRow);
			aux = new ArrayList<>(data);
			// Now you can merge-sort this data
		} catch (IOException e) {
			System.err.println("Error reading file: " + e.getMessage());
		}
		
		if (data != null ) {

			// System.out.println("ID" + data.get(2).id + ", Value: " + data.get(2).value);
			// data = reader.MergeSort(data);
			merge_sort_step sorter = new merge_sort_step();
			List<List<DataPair>> snapshots = new ArrayList<>();
			// data = sorter.MergeSortWithSteps(data, snapshots);
			snapshots.add(new ArrayList<>(data));
			sorter.mergeSortInPlace(data, aux, 0, data.size(), snapshots);
			String filename = "merge_sort_step_" + startingRow + "_" + endingRow + ".txt";
			try (BufferedWriter writer = new BufferedWriter(new FileWriter(filename))) {
				for (List<DataPair> snapshot : snapshots) {
					writer.write("[");
					for (int i = 0; i < snapshot.size(); i++) {
						DataPair pair = snapshot.get(i);
						writer.write(pair.id + "/" + pair.value);
						if (i != snapshot.size() - 1) {
							writer.write(", ");
						}
					}
					writer.write("]");
					writer.newLine();
				}
			} catch (IOException e) {
				System.err.println("Error writing snapshots: " + e.getMessage());
			}
		} else {
			System.out.println("No data to display.");
		}
	}
	
}

File: merge_sort_step.py
import time
import csv

# Divide and Conquer Process
def merge_sort(data, aux, left, right, snapshots):

    if right - left <= 1:
        return
    # Calculate the mid-point
    mid = (left + right + 1) // 2
    # recursive call
    merge_sort(data, aux, left, mid, snapshots) #recursive call on left side
    merge_sort(data, aux, mid, right, snapshots) #recursive call on right side

    # return merge(left, right)
    return merge_and_snapshot(data, aux, left, mid, right, snapshots)

def merge_and_snapshot(data, aux, left, mid, right, snapshots):
    i, j, k = left, mid, left
    while i < mid and j < right:
        if data[i][0] <= data[j][0]:
            aux[k] = data[i]
            i += 1
        else:
            aux[k] = data[j]
            j += 1
        k += 1

    while i < mid:
        aux[k] = data[i]
        i += 1
        k += 1

    while j < right:
        aux[k] = data[j]
        j += 1
        k += 1

    for t in range(left, right):
        data[t] = aux[t]
    # Take a snapshot of the current state of the data
    snapshots.append(list(data))

def merge(data, aux, left, right):
    result = []
    i = j = 0

    while i < len(left) and j < len(right):
        if left[i][0] <= right[j][0]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1

    result.extend(left[i:])
    result.extend(right[j:])
    for item in result:
        print(f"Sorted Item: {item[0]}, {item[1]}")
    print(f"Total Sorted Items: {len(result)}")
    return result

def get_input_dataset_from_file(filename, startingRow, endingRow):
    dataset = []
    try:
        with open(filename, 'r') as file:
            # i = startingRow
            i = 0
            # Skip lines until the starting row
            while i < startingRow-1:
                next(file)
                i += 1
            # Read the dataset from the file
            for line in file:
                if i >= endingRow and endingRow > 0:
                    break
                if ',' not in line:
                    continue 
                i += 1
                number_str, text = line.strip().split(',')
                number = int(number_str)
                dataset.append((number, text))
    except FileNotFoundError:
        print(f"File '{filename}' not found.")
        return []
    except ValueError:
        print("File contains invalid data. Ensure format is 'number,string' per line.")
        return []
    return dataset

def write_sorted_output_to_csv(sorted_data, startingRow, endingRow):
    output_filename = f"python_merge_sort_Step_{startingRow}_{endingRow}.txt"
    try:
        with open(output_filename, mode='w', newline='') as file:
            # writer = csv.writer(file)
            for data in sorted_data:
                file.write("[")
                for number, text in data:
                    file.write(f"{number}/{text}, ")
                if data:
                    file.seek(file.tell() - 2)  # Move back to remove last ", "
                file.write("]\n")
        print(f"Sorted data successfully written to '{output_filename}'.")
    except Exception as e:
        print(f"Error writing to file: {e}")

# Main Process
if __name__ == "__main__":
    # Reading Input Parameters
    filename = input("Enter the dataset CSV filename: ").strip()
    startingRow = int(input("Enter the starting row number (1-based index): ").strip())
    if startingRow < 0:
        print("Starting row number must be 1 or greater. Exiting.")
        exit()
    endingRow = int(input("Enter the ending row number (1-based index, 0 for no limit): ").strip())
    if endingRow < 0:
        print("Ending row number must be 0 or greater. Exiting.")
        exit()
    dataset = []
    dataset = get_input_dataset_from_file(filename, startingRow, endingRow)
    aux = dataset.copy()
    # Creating smpty list to hold snapshotscls

    snapshots = []
    if not dataset:
        print("No valid data to sort. Exiting.")
    else:
        snapshots.append(list(dataset))
        start = time.time()
        # Ensure 'right' does not exceed the length of the dataset
        # right_bound = min(len(dataset), endingRow - startingRow + 1 if endingRow > 0 else len(dataset))
        right_bound = len(dataset)
        sorted_dataset = merge_sort(dataset, aux, 0, right_bound, snapshots)
        end = time.time()
        runtime = end - start

        if runtime > 60:
            mins = int(runtime // 60)
            secs = runtime % 60
            print(f"\nTotal Sorting Runtime: {mins} minutes {secs:.2f} seconds")
        else:
            print(f"\nTotal Sorting Runtime: {runtime:.7f} seconds")

        write_sorted_output_to_csv(snapshots, startingRow, endingRow)
